File: packages/protocol/src/actor-compatibility.ts
import type { Actor, ActorRef, AgentActor, AgentProvenance } from '@agent-room/protocol-types';

export type CompatibleActor = Actor | ActorRef;

function isCurrentActor(actor: CompatibleActor): actor is Actor {
  const kind: unknown = Reflect.get(actor, 'kind');
  return kind === 'human' || kind === 'agent';
}

function projectLegacyProvenance(provenance: ActorRef['provenance']): AgentProvenance {
  return provenance === 'autonomous_agent' ? 'autonomous_agent' : 'human_confirmed_agent';
}

/**
 * 把 v1 中强制伪装成 Agent 的主体投影为 v2 判别联合。
 *
 * v1 的 `human` 仍然携带 Agent/实例身份，无法无损恢复 Principal；因此只能保守投影为
 * `human_confirmed_agent`，绝不能捏造一个 HumanActor。
 */
export function projectCompatibleActor(actor: CompatibleActor): Actor {
  if (isCurrentActor(actor)) {
    return actor;
  }

  const projected: AgentActor = {
    ...actor,
    kind: 'agent',
    provenance: projectLegacyProvenance(actor.provenance),
  };
  return Object.freeze(projected);
}

File: crates/protocol-conformance/src/actor_compatibility.rs
use crate::generated::{Actor, ActorRef, AgentActor, AgentProvenance, Provenance};

/// 把 v1 的无判别 Agent 主体投影成 v2 的显式 Agent 主体。
///
/// v1 的 `human` 没有 Principal 标识，不能安全伪造 `HumanActor`，因此保守映射为
/// `human_confirmed_agent`。
#[must_use]
pub fn project_legacy_actor(actor: ActorRef) -> Actor {
    let provenance = match actor.provenance {
        Provenance::AutonomousAgent => AgentProvenance::AutonomousAgent,
        Provenance::Human | Provenance::HumanConfirmedAgent => AgentProvenance::HumanConfirmedAgent,
    };

    Actor::Agent(AgentActor {
        agent: actor.agent,
        instance_id: actor.instance_id,
        kind: "agent".to_owned(),
        provenance,
        extensions: actor.extensions,
    })
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use crate::generated::{ActorRef, AgentProvenance, AgentRef, Provenance};

    use super::{Actor, project_legacy_actor};

    #[test]
    fn 旧人类来源不会被伪造成缺少主体标识的人类() {
        let projected = project_legacy_actor(ActorRef {
            agent: AgentRef {
                agent_id: "01945c1e-7b5a-7c7f-8a28-2de53f56a9a3".to_owned(),
                avatar_url: None,
                display_name: "构建助手".to_owned(),
                matrix_user_id: "@build-agent:agent-room.test".to_owned(),
                extensions: BTreeMap::new(),
            },
            instance_id: "01945c1e-7b5a-7c7f-8a28-2de53f56a9a4".to_owned(),
            provenance: Provenance::Human,
            extensions: BTreeMap::new(),
        });

        let Actor::Agent(actor) = projected else {
            panic!("v1 Agent 主体只能投影成 AgentActor");
        };
        assert_eq!(actor.kind, "agent");
        assert_eq!(actor.provenance, AgentProvenance::HumanConfirmedAgent);
    }
}
